src/cmd/ksh93/sh/io.c

```diff
@@ -1226,7 +1226,10 @@ int sh_redirect(Shell_t *shp, struct ionod *iop, int flag) {
                         message = e_file;
                         goto fail;
                     }
-                    sh_iovalidfd(shp, dupfd);
+                    if (!sh_iovalidfd(shp, dupfd)) {
+                        message = e_file;
+                        goto fail;
+                    }
```

Comment:
This looks fine but I'm suspicious of every single place that calls `sh_iovalidfd()` without checking it's return value. It seems to me that at a minimum if a failure should be a "can't happen" situation those sites should be rewritten as follows if we can't handle it as a soft failure like you're doing with this change:

```
if (!sh_iovalidfd(shp, dupfd)) abort();
```

Reply:
I have opened #595 to track this.